client/changelist/changelist.go

```diff
@@ -32,3 +36,30 @@ func (cl *memChangelist) Clear(archive string) error {
 func (cl *memChangelist) Close() error {
 	return nil
 }
+
+func (cl *memChangelist) NewIterator() (ChangeIterator, error) {
+	return &MemChangeListIterator{index: 0, collection: cl.changes}, nil
+}
+
+// MemChangeListIterator is a concrete instance of ChangeIterator
+type MemChangeListIterator struct {
+	index      int
+	collection []Change // Same type as memChangeList.changes
+}
+
+// Next returns the next Change
+func (m *MemChangeListIterator) Next() (item Change, err error) {
+	defer func() {
+		if e := recover(); e != nil {
+			err = fmt.Errorf("Error in Next(): %v ", e)
+		}
+	}()
```

Comment:
This seems like a convoluted way of doing a bounds check. I think it would be much better to explicitly check the slice bounds with `len`.